src/node/txpackagetracker.cpp

```diff
@@ -4,41 +4,157 @@
-    bool OrphanageAddTx(const CTransactionRef& tx, NodeId peer) { return m_orphanage.AddTx(tx, peer); }
     bool OrphanageHaveTx(const GenTxid& gtxid) { return m_orphanage.HaveTx(gtxid); }
     CTransactionRef GetTxToReconsider(NodeId peer) { return m_orphanage.GetTxToReconsider(peer); }
-    void DisconnectedPeer(NodeId peer) {
-        m_orphanage.EraseForPeer(peer);
+    void DisconnectedPeer(NodeId nodeid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        orphan_request_tracker.DisconnectedPeer(nodeid);
+        m_orphanage.EraseForPeer(nodeid);
+    }
+    void BlockConnected(const CBlock& block) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        const auto wtxids_erased{m_orphanage.EraseForBlock(block)};
+        std::set<uint256> block_wtxids;
+        std::set<uint256> conflicted_wtxids;
+        for (const CTransactionRef& ptx : block.vtx) {
+            block_wtxids.insert(ptx->GetWitnessHash());
+        }
+        for (const auto& wtxid : wtxids_erased) {
+            if (block_wtxids.count(wtxid) == 0) {
+                conflicted_wtxids.insert(wtxid);
+            }
+        }
+        for (const auto& wtxid : block_wtxids) {
+            orphan_request_tracker.ForgetTxHash(wtxid);
+        }
+        for (const auto& wtxid : conflicted_wtxids) {
+            orphan_request_tracker.ForgetTxHash(wtxid);
+        }
     }
-    void BlockConnected(const CBlock& block) { m_orphanage.EraseForBlock(block); }
     void LimitOrphans(unsigned int max_orphans) { m_orphanage.LimitOrphans(max_orphans); }
     bool HaveTxToReconsider(NodeId peer) { return m_orphanage.HaveTxToReconsider(peer); }
     size_t OrphanageSize() { return m_orphanage.Size(); }
-    void MempoolAcceptedTx(const CTransactionRef& ptx)
+    void MempoolAcceptedTx(const CTransactionRef& ptx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
     {
+        LOCK(m_mutex);
         m_orphanage.AddChildrenToWorkSet(*ptx);
         m_orphanage.EraseTx(ptx->GetWitnessHash());
     }
-    void MempoolRejectedTx(const uint256& wtxid)
+    void MempoolRejectedTx(const uint256& wtxid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
     {
+        LOCK(m_mutex);
         m_orphanage.EraseTx(wtxid);
     }
+    void AddOrphanTx(NodeId nodeid, const uint256& wtxid, const CTransactionRef& tx, bool is_preferred, std::chrono::microseconds reqtime)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        // Skip if we weren't provided the tx and can't find the wtxid in the orphanage.
+        if (tx == nullptr && !m_orphanage.HaveTx(GenTxid::Wtxid(wtxid))) return;
+
+        // Even though this stores the orphan wtxid, GenTxid::Txid instead of Wtxid because we will be requesting the parents via txid.
+        orphan_request_tracker.ReceivedInv(nodeid, GenTxid::Txid(wtxid), is_preferred, reqtime);
```

Comment:
(moved over from [27742](https://github.com/bitcoin/bitcoin/pull/27742#discussion_r1265739284)):
The approach to add a wtxid disguised as a txid to distinguish between legacy orphan processing and package relay seems a bit like a hack to me.
I'm not strictly against it, but I guess I don't completely understand yet why it's necessary - the information whether a peer supports package relay does not change, so why can't we just always use `GenTxid::Wtxid(wtxid)` and check again in `GetOrphanRequests()` and elsewhere whether we do package relay with the peer instead of checking whether it's a txid / wtxid?

Reply:
Ah good point, it's not necessary, we could just use `GenTxid::Wtxid` for all of them. :+1: